modules/rate-limiting/v2/ibc_middleware.go

```diff
@@ -91,23 +108,45 @@ func (im IBCMiddleware) OnAcknowledgementPacket(
 	payload channeltypesv2.Payload,
 	relayer sdk.AccAddress,
 ) error {
-	packet := v2ToV1Packet(payload, sourceClient, destinationClient, sequence)
+	packet, err := v2ToV1Packet(payload, sourceClient, destinationClient, sequence)
+	if err != nil {
+		im.keeper.Logger(ctx).Error(fmt.Sprintf("ICS20 rate limiting OnAckPacketfailed to convert v2 packet to v1 packet: %s", err.Error()))
+		return err
+	}
 	if err := im.keeper.AcknowledgeRateLimitedPacket(ctx, packet, acknowledgement); err != nil {
 		im.keeper.Logger(ctx).Error(fmt.Sprintf("ICS20 RateLimited OnAckPacket failed: %s", err.Error()))
 		return err
 	}
 	return im.app.OnAcknowledgementPacket(ctx, sourceClient, destinationClient, sequence, acknowledgement, payload, relayer)
 }
 
-func v2ToV1Packet(payload channeltypesv2.Payload, sourceClient, destinationClient string, sequence uint64) channeltypes.Packet {
+func v2ToV1Packet(payload channeltypesv2.Payload, sourceClient, destinationClient string, sequence uint64) (channeltypes.Packet, error) {
+	transferRepresentation, err := transfertypes.UnmarshalPacketData(payload.Value, payload.Version, payload.Encoding)
+	if err != nil {
+		return channeltypes.Packet{}, err
+	}
+
+	packetData := transfertypes.FungibleTokenPacketData{
+		Denom:    transferRepresentation.Token.Denom.Path(),
+		Amount:   transferRepresentation.Token.Amount,
+		Sender:   transferRepresentation.Sender,
+		Receiver: transferRepresentation.Receiver,
+		Memo:     transferRepresentation.Memo,
+	}
+
+	packetDataBz, err := json.Marshal(packetData)
```

Comment:
Note that the keeper expects a json marshaled `FungibleTokenPacketData`. I am choosing to make v2 do the unmarshal and remarshal rather than change the keeper logic, so that if there is an issue here, ibc classic will not be affected.

Reply:
Same thing we did in callbacks too